daemon: move directory setup into `SetUpTest`

ianvernon opened the pull request:
Setting up the directories and associated configuration per-test means
that there is less global state to worry about that may leak between
tests if the configuration was set up in `SetUpMain`. Factor out this
setup into a function for better readability and move it into `SetUpTest`.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9183)
<!-- Reviewable:end -->

Commit:
daemon: move directory setup into `SetUpTest`

Setting up the directories and associated configuration per-test means
that there is less global state to worry about that may leak between
tests if the configuration was set up in `SetUpMain`. Factor out this
setup into a function for better readability and move it into `SetUpTest`.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on daemon/daemon_test.go:
```diff
@@ -86,6 +86,9 @@ func TestMain(m *testing.M) {
 		panic("TempDir() failed.")
 	}
+
+	defer os.RemoveAll(tempRunDir)
```
But AFAIK, TestMain doesn't block until the tests are run.

ianvernon replied:
I don't understand what you mean here :/ are you saying that this won't run because we run `ox.Exit(m.Run()`?

ianvernon replied:
Code has changed, PTAL

aanm commented on daemon/daemon_test.go:
```diff
@@ -149,10 +152,6 @@ func (ds *DaemonSuite) SetUpTest(c *C) {
 	ds.d.endpointManager.RemoveAll()
 
-	if ds.d != nil {
-		os.RemoveAll(option.Config.RunDir)
```
Shouldn't the `TearUpTest` create a new `option.Config.RunDir` as well as leaving this change untouched?

ianvernon replied:
I'll look into seeing whether the logic can be put into `SetUpTest`. I definitely think that we should not have the `ds.d != nil` check, though, and just unconditionally try to remove the existing `option/config/RunDir`, as we create the directories regardless of if the Daemon is nil or not.

ianvernon replied:
Refactored the configuration setup to be ran before each test. I had to make a change to ensure that metrics registration was only done once per run of the tests within `option.Config.Populate()`.

aanm commented on pkg/option/config.go:
```diff
@@ -1762,7 +1763,12 @@ func (c *DaemonConfig) Populate() {
 	c.MetricsConfig, collectors = metrics.CreateConfiguration(metricsSlice)
-	metrics.MustRegister(collectors...)
+
+	// Only run this once so we can call this function multiple times in unit
```
`TestMain` was added for this reason :) cc @jrajahalme

ianvernon replied:
I can do the following in `TestMain` and revert the changes to `pkg/option/config.go`:

```
option.Config.Populate()	 	
option.Config.IdentityAllocationMode = option.IdentityAllocationModeKVstore
```

Would this alleviate your concerns? What is blocking here?

jrajahalme replied:
I'd prefer keeping the TestMain() for the calls that must be done only once.

ianvernon replied:
Got approval from Jarno, PTAL

jrajahalme commented on daemon/daemon_test.go:
```diff
@@ -149,9 +156,7 @@ func (ds *DaemonSuite) SetUpTest(c *C) {
-	if ds.d != nil {
-		os.RemoveAll(option.Config.RunDir)
-	}
+	os.RemoveAll(option.Config.RunDir)
```
Is this done also when the test fails? It would be nice to have the temp files around when a test fails.

ianvernon replied:
Done, PTAL

Commit:
daemon: do not delete directories created by tests if tests fail

This allows for easier debugging of tests (see https://github.com/cilium/cilium/pull/9183#discussion_r323956613).

Signed-off by: Ian Vernon <ian@cilium.io>
